src/v/serde/protobuf/tests/data_generator.cc

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "serde/protobuf/tests/data_generator.h"
+
+#include "random/generators.h"
+
+#include <cstdint>
+#include <memory>
+
+namespace testing {
+ss::sstring random_string(const protobuf_generator_config& config) {
+    auto [min, max] = config.string_length_range;
+    return random_generators::gen_alphanum_string(
+      random_generators::get_int(min, max));
+}
+
+std::unique_ptr<google::protobuf::Message>
+protobuf_generator::generate_protobuf_message_impl(
+  int level, const google::protobuf::Descriptor* d) {
+    auto& factory = _factory;
+    auto get_elements_count = [level, this]() -> size_t {
+        if (level >= _config.max_nesting_level) {
+            return 0;
+        }
+        return _config.elements_in_collection.value_or(
+          random_generators::get_int<size_t>(10));
+    };
+
+    auto message = std::unique_ptr<google::protobuf::Message>(
+      factory.GetPrototype(d)->New());
+    const google::protobuf::Reflection* r = message->GetReflection();
+    for (int i = 0; i < d->field_count(); i++) {
+        const auto* field_d = d->field(i);
+
+        // if its a union type select a random field in it, randomize its value,
+        // then jump to the end of all fields in the oneof.
+        if (auto oneof_d = field_d->real_containing_oneof()) {
+            auto num_fields = oneof_d->field_count();
+            field_d = d->field(
+              random_generators::get_int(i, i + (num_fields - 1)));
+            i += (num_fields - 1);
+
+            if (
+              _config.randomize_optional_fields && field_d->is_optional()
+              && random_generators::get_int(0, 1) == 1) {
+                continue;
+            }
+        }
+
+        using fdns = google::protobuf::FieldDescriptor;
+
+        if (field_d->is_repeated()) {
+            for (size_t i = 0; i < get_elements_count(); i++) {
+                switch (field_d->cpp_type()) {
```

Comment:
any reason we are using `cpp_type` but not `type` ?

Reply:
Its a smaller set of types, and for the purpose of randomizing a message we only really care what C++ type protobuf wants the field set to.